sdk/src/azure/core/az_mqtt5_rpc.c

```diff
@@ -0,0 +1,191 @@
+  {
+    ret = false;
+  }
+#else // TRANSPORT_MOSQUITTO
+  (void)sub;
+  (void)topic;
+  ret = false;
+#endif
+  return ret;
+}
+
+AZ_NODISCARD bool az_mqtt5_rpc_status_failed(az_mqtt5_rpc_status status)
+{
+  return (status < 200 || status >= 300);
+}
+
+AZ_NODISCARD az_result az_rpc_get_topic_from_format(
+    az_span model_id,
+    az_span executor_client_id,
+    az_span invoker_client_id,
+    az_span command_name,
+    az_span format,
+    az_span out_topic,
+    int32_t* out_topic_length)
+{
+  const az_span model_id_key = AZ_SPAN_FROM_STR("{serviceId}");
+  const az_span command_name_key = AZ_SPAN_FROM_STR("{name}");
+  const az_span executor_id_key = AZ_SPAN_FROM_STR("{executorId}");
+  const az_span invoker_id_key = AZ_SPAN_FROM_STR("{invokerId}");
+
+  int32_t format_size = az_span_size(format);
+
+  // Determine the length of the final topic
+  int32_t topic_length = format_size;
+  bool has_substitution = false;
+  int32_t max_temp_length = 0;
+  if (az_span_find(format, model_id_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(model_id, 1, false);
+    topic_length += az_span_size(model_id);
+    topic_length -= az_span_size(model_id_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+  if (az_span_find(format, command_name_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(command_name, 1, false);
+    topic_length += az_span_size(command_name);
+    topic_length -= az_span_size(command_name_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+  if (az_span_find(format, executor_id_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(executor_client_id, 1, false);
+    topic_length += az_span_size(executor_client_id);
+    topic_length -= az_span_size(executor_id_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+  if (az_span_find(format, invoker_id_key) >= 0)
+  {
+    has_substitution = true;
+    _az_PRECONDITION_VALID_SPAN(invoker_client_id, 1, false);
+    topic_length += az_span_size(invoker_client_id);
+    topic_length -= az_span_size(invoker_id_key);
+    max_temp_length = topic_length > max_temp_length ? topic_length : max_temp_length;
+  }
+
+  // Must be large enough to fit the entire format as items are substituted throughout the function
+  // even if that's larger than the output topic
+  _az_PRECONDITION_VALID_SPAN(out_topic, max_temp_length, true);
+
+  // if there are no substitutions, just copy the format to the output
+  if (!has_substitution)
+  {
+    az_span_copy(out_topic, format);
+    if (out_topic_length != NULL)
+    {
+      *out_topic_length = format_size;
+    }
+    return AZ_OK;
+  }
+
+  // Must be large enough to fit the entire format even if that's larger than the output topic
```

Comment:
Suggestion:
create a _az_span_replace(az_span in, az_span from, az_span to, az_span* out_replaced);
and shorten the code below.